### PR TITLE
Fail SHOW queries on objects that are not accessible to user

### PR DESCRIPTION
**Pull request by kokosing**
Fail SHOW queries on objects that are not accessible to user

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/metadata/QualifiedObjectName.java [hunk 1]
```diff
@@ -85,6 +86,11 @@ public QualifiedTablePrefix asQualifiedTablePrefix()
         return new QualifiedTablePrefix(catalogName, schemaName, objectName);
     }
 
+    public CatalogSchemaName asCatalogSchemaName()
+    {
+        return new CatalogSchemaName(catalogName, schemaName);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/metadata/QualifiedObjectName.java]: This doesn't return "another way to refer to `this`", so method should not be named `as*`. Rename to `getCatalogSchemaName`

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 2]
```diff
@@ -140,6 +142,7 @@ public Statement rewrite(
         return (Statement) new Visitor(metadata, parser, session, parameters, accessControl).process(node, null);
     }
 
+    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: Don't suppress on the class level, unless the suppression applies to the whole class.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 3]
```diff
@@ -370,14 +371,20 @@ protected Node visitShowCreate(ShowCreate node, Void context)
...
                 if (!viewDefinition.isPresent()) {
-                    if (metadata.getTableHandle(session, objectName).isPresent()) {
+                    if (isTableAccessible(objectName)) {
                         throw new SemanticException(NOT_SUPPORTED, node, "Relation '%s' is a table, not a view", objectName);
                     }
                     throw new SemanticException(MISSING_TABLE, node, "View '%s' does not exist", objectName);
                 }
 
+                if (!isViewAccessible(objectName)) {
+                    throw new SemanticException(NOT_SUPPORTED, node, "View '%s' does not exist", objectName);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: new code path? new test?

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 4]
```diff
@@ -558,6 +566,51 @@ protected Node visitShowSession(ShowSession node, Void context)
...
+
+            if (!isViewAccessible(objectName) && !isTableAccessible(objectName)) {
+                throw new SemanticException(MISSING_TABLE, node, "Table '%s' does not exist", objectName);
+            }
+        }
+
+        private boolean isCatalogAccessible(String catalogName)
+        {
+            return accessControl.filterCatalogs(session.getIdentity(), ImmutableSet.of(catalogName)).contains(catalogName)
+                    && metadata.catalogExists(session, catalogName);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: i would reverse -- check access control only for existing catalogs. I think an access control impl may have problems for non-existing catalogs.

(here & below)

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 5]
```diff
@@ -558,6 +566,51 @@ protected Node visitShowSession(ShowSession node, Void context)
...
+        {
+            if (!isCatalogAccessible(objectName.getCatalogName())) {
+                throw new SemanticException(MISSING_SCHEMA, node, "Catalog '%s' does not exist", objectName.getCatalogName());
+            }
+
+            if (!isSchemaAccessible(objectName.asCatalogSchemaName())) {
+                throw new SemanticException(MISSING_SCHEMA, node, "Schema '%s' does not exist", objectName.getSchemaName());
+            }
+
+            if (!isViewAccessible(objectName) && !isTableAccessible(objectName)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: Why isn't this `if (!isRelationAccessible(objectName))` ?
(also, sometimes we say "table" meaning "table or view" (`accessControl.filterTables` ?), so `isTableAccessible` could accept views... but i am not a fan of this approach)

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 6]
```diff
@@ -558,6 +566,51 @@ protected Node visitShowSession(ShowSession node, Void context)
...
+            schemas = accessControl.filterSchemas(session.getRequiredTransactionId(), session.getIdentity(), schema.getCatalogName(), schemas);
+            return schemas.contains(schema.getSchemaName()) && metadata.schemaExists(session, schema);
+        }
+
+        private boolean isTableAccessible(QualifiedObjectName tableName)
+        {
+            return isTableOrViewVisibleToUser(tableName) && metadata.getTableHandle(session, tableName).isPresent();
+        }
+
+        private boolean isViewAccessible(QualifiedObjectName tableName)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: `viewName`

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 7]
```diff
@@ -558,6 +566,51 @@ protected Node visitShowSession(ShowSession node, Void context)
...
+        {
+            return isTableOrViewVisibleToUser(tableName) && metadata.getView(session, tableName).isPresent();
+        }
+
+        private boolean isTableOrViewVisibleToUser(QualifiedObjectName tableName)
+        {
+            Set<SchemaTableName> tableNames = ImmutableSet.of(tableName.asSchemaTableName());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: inline

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 8]
```diff
@@ -558,6 +566,51 @@ protected Node visitShowSession(ShowSession node, Void context)
...
+
+        private boolean isCatalogAccessible(String catalogName)
+        {
+            return accessControl.filterCatalogs(session.getIdentity(), ImmutableSet.of(catalogName)).contains(catalogName)
+                    && metadata.catalogExists(session, catalogName);
+        }
+
+        private boolean isSchemaAccessible(CatalogSchemaName schema)
+        {
+            Set<String> schemas = ImmutableSet.of(schema.getSchemaName());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: inline

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java [hunk 9]
```diff
@@ -370,14 +371,20 @@ protected Node visitShowCreate(ShowCreate node, Void context)
             QualifiedObjectName objectName = createQualifiedObjectName(session, node, node.getName());
             Optional<ViewDefinition> viewDefinition = metadata.getView(session, objectName);
 
+            checkObjectAccessible(node, objectName);
+
             if (node.getType() == VIEW) {
                 if (!viewDefinition.isPresent()) {
-                    if (metadata.getTableHandle(session, objectName).isPresent()) {
+                    if (isTableAccessible(objectName)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: didn't `checkObjectAccessible` above do that already?

File: presto-main/src/main/java/com/facebook/presto/metadata/QualifiedObjectName.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/metadata/QualifiedObjectName.java]: I fixed that. Thanks

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 3 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: This check is redundant. I am going to remove it.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java (hunk 5 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/rewrite/ShowQueriesRewrite.java]: The best would be to separate them and have `filterViews`

**Comment by kokosing** [comment]
> I just skimmed. No need for new tests? I think the behavior is being changed here.

Today, when roles are not yet merged there is not much usage of `filter*` methods. So technically there is no behaviour change.

**Pull request closed without merging**